src/Routing/PushState.purs

```diff
@@ -0,0 +1,247 @@
+      pure $ modifyRef listenersRef $ StrMap.delete (show fresh)
+
+    locationState = do
+      loc <- DOM.window >>= DOM.location
+      state <- DOM.window >>= DOM.history >>= DOM.state
+      pathname <- fromMaybe <*> String.stripPrefix (String.Pattern (unwrap basename)) <$> DOM.pathname loc
+      search <- DOM.search loc
+      hash <- DOM.hash loc
+      let path = pathname <> search <> hash
+      pure { state, pathname, search, hash, path }
+
+  -- The hashchange interface is asynchronous, since hashchange events are
+  -- fired on the next tick of the event loop. We want the push-state
+  -- interface to behave as similarly as possible, so we use the microtask
+  -- queue via MutationObserver to schedule callbacks. Alternatively we could
+  -- just use a setTimeout, but it would not be as prompt. We use a fresh
+  -- counter so that the text change mutation always fires.
+  schedule <- do
+    obsvNode <-
+      DOM.window
+        >>= DOM.document
+        >>> map DOM.htmlDocumentToDocument
+        >>= DOM.createTextNode ""
+        >>> map DOM.textToNode
+    observer <- DOM.mutationObserver \_ _ -> do
+      argsQueue <- readRef argsRef
+      writeRef argsRef []
+      for_ argsQueue notify
+    DOM.observe obsvNode { characterData: true } observer
+    pure \args -> do
+      argsQueue <- readRef argsRef
+      writeRef argsRef (Array.snoc argsQueue args)
+      when (Array.null argsQueue) do
+        fresh ← readRef freshRef
+        writeRef freshRef (fresh + 1)
+        DOM.setNodeValue (show fresh) obsvNode
+
+  let
+    stateFn op state path = do
+      let
+        loc = normalizeLocation state path
+        url = DOM.URL $ unwrap basename <> loc.pathname <> loc.search <> loc.hash
+      DOM.window
+        >>= DOM.history
+        >>= op state (DOM.DocumentTitle "") url
+      schedule loc
+
+    listener =
+      DOM.eventListener \_ -> notify =<< locationState
+
+  DOM.window
+    >>= DOM.windowToEventTarget
+    >>> DOM.addEventListener DOM.popstate listener false
+
+  pure
+    { pushState: stateFn DOM.pushState
+    , replaceState: stateFn DOM.replaceState
+    , locationState
+    , listen
+    }
+
+type LocationState =
+  { state :: Foreign
+  , path :: String
+  , pathname :: String
+  , search :: String
+  , hash :: String
+  }
+
+normalizeLocation :: Foreign -> Path -> LocationState
+normalizeLocation state (Path path) =
+  case searchIx, hashIx of
+    Nothing, Nothing   -> { state, pathname: path, hash: "", search: "", path }
+    Nothing, Just hix  -> { state, pathname: String.take hix path, search: "", hash: String.drop hix path, path }
+    Just six, Nothing  -> { state, pathname: String.take six path, search: String.drop six path, hash: "", path }
+    Just six, Just hix -> { state, pathname: String.take six path, search: String.take (hix - six) (String.drop six path), hash: String.drop hix path, path }
+  where
+  searchIx = String.indexOf (String.Pattern "?") path
```

Comment:
`purescript-uri` for parsing here?

Reply:
I think it's overkill. I would be parsing it and then having to just flatten it out again because the structure returned by purescript-uri is too "rich". I'm trying to reflect the properties provided by the actual DOM.

Reply:
Also next version of ps-uri will be even more overkill then current one

Reply:
Okay